src/borrow_tracker/tree_borrows/tree.rs

```diff
@@ -725,27 +736,480 @@ mod commutation_tests {
+            Ok(())
+        }
+    }
+
+    #[derive(Clone, PartialEq, Eq, Hash)]
+    /// A `(LocStateProt, LocStateProt)` has more degrees of freedom
+    /// than what we want to allow for two aliasing and mutually foreign
+    /// tags (e.g. `(Active, Active)` among many others is unreachable through
+    /// normal means), so a pair of tags gets its own methods.
+    /// We'll still generate these cases exhaustively, but it prevents silly mistakes.
+    /// Pointers `x` and `y` might be used in different threads, so their protectors
+    /// could be removed in an arbitrary order.
+    struct LocStateProt2 {
+        x: LocStateProt,
+        y: LocStateProt,
+    }
+
+    impl LocStateProt2 {
+        fn perform_access2(&self, acc: Access2) -> Result<Self, ()> {
+            use AccessRel2 as Rel2;
+            use AccessRelatedness as Rel1;
+            // We waited until the last moment to convert
+            // `AccessRel2` into `(AccessRelatedness, AccessRelatedness)`
+            // because not all combinations can occur.
+            let xrel = match acc.rel2 {
+                Rel2::ChildX => Rel1::StrictChildAccess,
+                _ => Rel1::DistantAccess,
+            };
+            let yrel = match acc.rel2 {
+                Rel2::ChildY => Rel1::StrictChildAccess,
+                _ => Rel1::DistantAccess,
+            };
+            let x = self.x.perform_access(acc.kind, xrel)?;
+            let y = self.y.perform_access(acc.kind, yrel)?;
+            Ok(Self { x, y })
+        }
+
+        /// End of protector on x.
+        fn release_x(&self) -> Result<Self, ()> {
+            let x = self.x.release();
+            let y = self.y.clone();
+            Ok(Self { x, y })
+        }
+
+        /// End of protector on y.
+        fn release_y(&self) -> Result<Self, ()> {
+            let x = self.x.clone();
+            let y = self.y.release();
+            Ok(Self { x, y })
+        }
+    }
+
+    impl Exhaustive for LocStateProt2 {
+        fn exhaustive() -> Box<dyn Iterator<Item = Self>> {
+            Box::new(<[LocStateProt; 2]>::exhaustive().map(|[x, y]| Self { x, y }))
+        }
+    }
+
+    impl fmt::Display for LocStateProt2 {
+        fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+            write!(f, "x:{}, y:{}", self.x, self.y)
+        }
+    }
+
+    // Arbitrary sequence of accesses, as experienced by two unrelated pointers.
+    #[derive(Debug, Clone)]
+    struct Opaque2 {
+        events: Vec<Option<Access2>>,
+    }
+
+    impl Exhaustive for Opaque2 {
+        fn exhaustive() -> Box<dyn Iterator<Item = Self>> {
+            // There are a lot of combinations (infinitely many in fact),
+            // so we're only generating a small subset and then testing through other means
+            // that we have a representative sample.
+            // The skips that we introduce here take advantage of
+            // - beyond length 4 there is a shorter sequence that has the same effect
```

Comment:
Why?

Also the length you are suing in the end is 3, not 4. But the comment here indicates 4 ("3 or less" can be None, "more than 4" is not needed -> 4 is what we want)